Document Ganache CLI options (#1312)
trufflesuite/trufflesuite.com: src/docs/ganache/reference/cli-options.md

This reference describes the syntax of the Ganache command line interface (CLI) options for the Ethereum
network.

## Specify options

The startup options are grouped in the `chain`, `database`, `fork`, `logging`, `miner`, and `wallet` namespaces,
and should be used as follows on startup:

```bash
ganache --<namespace>.<option>="<value>"
```

You can also use an alias for startup options that have them. For example to set the network ID of a
network use `ganache -i=<NETWORK_ID>` instead of `ganache --chain.networkId=<NETWORK_ID>`.

When using the startup options programmatically, use the following:

```javascript
const options = { <namespace>: { <option>: <"value">}};
const provider = ganache.provider(options);
```

<p class="alert alert-warning">
<strong>Note</strong>: Command line options are case-sensitive.
</p>

## `chain`

**`--chain.allowUnlimitedContractSize`**

=== "Syntax"

    ```bash
    --chain.allowUnlimitedContractSize=<BOOLEAN>
    ```

=== "Example"

    ```bash
    --chain.allowUnlimitedContractSize=true
    ```

Allows unlimited contract sizes while debugging. Setting this to `true` causes Ganache to behave
differently than production environments.

> The default is `false`.

Author: Good catch. I've made these consistent.

** `--chain.asyncRequestProcessing`**

=== "Syntax"

    ```bash
    --chain.asyncRequestProcessing=<BOOLEAN>
    ```

=== "Example"

    ```bash
    --chain.asyncRequestProcessing=false
    ```

When set to `false`, only one request is processed at a time. The default is `true`.

**`--chain.chainId`**

=== "Syntax"

    ```bash
    --chain.chainId=<NUMBER>
    ```

=== "Example"

    ```bash
    --chain.chainId=10
    ```

Set the Chain ID of the network. The default is `1337`.

**`--chain.networkId`**, **`-i`**

=== "Syntax"

    ```bash
    --chain.networkId=<NUMBER>
    ```

=== "Example"

    ```bash
    --chain.networkId=10
    ```

Set the network ID that's retrieved when running the [`net_version`](https://ethereum.org/en/developers/docs/apis/json-rpc/#net_version)
JSON-RPC method.
The default is the system time when the process starts, or network ID of the forked
blockchain if configured.

The alias `--networkId` has been deprecated, use `-i` instead.

**`--chain.time`**, **`-t`**

=== "Syntax"

    ```bash
    --chain.time=<DATE>
    ```

=== "Example"

    ```bash
    --chain.time=1662596414
    ```

> Date and time that the first block should start.

Author: I also was unsure about this. Could we say, "Date at which the first block should start. Accepts the date in [Unix time format](https://en.wikipedia.org/wiki/Unix_time)."

However the following command (using ISO 8601) also works for me, so I don't know what formats are acceptable.

 ganache --chain.time="2022-09-27T21:03:00+00:00"

> Date and time that the first block should start.

Author: @emilyJLin95 ^^

> Date and time that the first block should start.

Author: I updated this too:

> Date and time that the first block should start. Accepts the date format supported by the
> [JavaScript `Date()` constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date).
> For example Unix epoch or a string.

The alias `--time` has been deprecated, use `-t` instead.

**`--chain.hardfork`**, **`-k`**

=== "Syntax"

    ```bash
    --chain.hardfork=<STRING>
    ```

=== "Example"

    ```bash
    --chain.hardfork="arrowGlacier"
    ```

Set the hardfork rules for the EVM. Valid options are: `constantinople`, `byzantium`, `petersburg`,
`istanbul`, `muirGlacier`, `berlin`, `london`, `arrowGlacier`, and `grayGlacier`.

The default is `grayGlacier`.

The alias `--hardfork` has been deprecated, use `-k` instead.

> **`--chain.vmErrorsOnRPCResponse`

Reviewer: Looks like this line is missing the closing `**`

    --miner.defaultTransactionGasLimit=<STRING>
    ```

=== "Example"

    ```bash
    --miner.defaultTransactionGasLimit="16F30"
    ```

Set the default transaction gas limit in WEI. Set to `estimate` to use an estimate (slows down transaction execution by 40%+).
The defaults is `0x15f90`.

**`--miner.difficulty`**

=== "Syntax"
    ```bash
    --miner.difficulty=<STRING>
    ```

=== "Example"

    ```bash
    --miner.difficulty="0x2"
    ```

Set the block difficulty. The default is `0x1`.

**`--miner.callGasLimit`**

=== "Syntax"
    ```bash
    --miner.callGasLimit=<STRING>
    ```

=== "Example"

    ```bash
    --miner.callGasLimit="0x58af080"
    ```

Set the transaction gas limit in WEI for `eth_call` and `eth_estimateGas` calls. The default is `0x2faf080`.

**`--miner.instamine`**

=== "Syntax"
    ```bash
    --miner.instamine=<STRING>
    ```

=== "Example"

    ```bash
    --miner.instamine="strict"
    ```

Set the `instamine` mode to either `eager` or `strict`. In `eager` mode a transaction is included in a block before its hash
is returned to the caller. In `strict` mode a transaction's hash is returned to the caller before the transaction is included in a block.

> This value has no effect if `--miner.blockTime` is *not* `0` (the default).`

Reviewer: Possible spurious backtick at the end of this line?

The default is `eager`.

**`--miner.coinbase`**

=== "Syntax"
    ```bash
    --miner.coinbase=<STRING>
    ```

=== "Example"

    ```bash
    --miner.coinbase="fe3b557e8fb62b89f4916b721be55ceb828dbd73"
    ```

Sets the mining reward address. The default is `0x0000000000000000000000000000000000000000`.

**`--miner.extraData`**

=== "Syntax"
    ```bash
    --miner.extraData=<STRING>
    ```

=== "Example"

    ```bash
    --miner.extraData="0x444F4E27542050414E4943202120484F444C2C20484F444C2C20484F444C2021"
    ```

> A hex string representing the 32 bytes included in the extra data field of a mined block. The default is `0x`

Reviewer: Missing full stop at the end of this paragraph.

**`--fork.password`**

=== "Syntax"
    ```bash
    --fork.password=<STRING>
    ```

=== "Example"

    ```bash
    --fork.password="changeme"
    ```

Password for basic authentication. Does not require setting `--fork.username`.
When combined with `--fork.username`, the shorthand is `{ headers: { "Authorization": "Basic {ENCODED-BASIC-HEADER}" } }`

If the `--fork.headers` specifies an authorization header, it will be inserted after the Basic token.

**`--fork.jwt`**

=== "Syntax"
    ```bash
    --fork.jwt=<STRING>
    ```

=== "Example"

    ```bash
    --fork.jwt="eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJwZXJtaXNzaW9ucyI6Iio6KiIsInByaXZhY3lQdWJsaWNLZXkiOiIyVUtIM1ZKVGhrT29Lc2tyTEZwd294Q25uUkFSeW9iVjFiRWRnc2VGSFRzPSIsImV4cCI6IjE2MDA4OTk5OTkwMDIiLCJpYXQiOjE2MzkxNTc2Mjd9.FGf-FmfDQlIPCRDGmNnsHZWlwrUr69d7AIDqQrIrUrSJLiwGpR3NCUhVHIDMpQvDHQYf-sFMZTYvZGrvztYRuBKWMbTfIZKN74onzNJbFIPBVQuUX2HMXmI4VQ3UFB11LShiUJHKLna13qdbqfbgJIO3HetxJhJQxTiwtixfHwyPXl-Nx8HbQy_AWH58lLAUeaoLzN7QIA9kborthBpvfK9C7Sv1lXT1cdCDC4oRKBoiMg2RWFZtGtxFsnWyloangwbhCB6Bc_elqY5nd9WkF4ix95xsP_HgBcouy1sDw6jxn5_LveX53H8owczVWP6S1e6hv6hq2fs6YkSntKMK2g"
    ```

> Encoded JSON Web Token (JWT) used for authenticating to servers.

Reviewer: Worth us adding a link out to https://jwt.io on the `JSON Web Token (JWT)` reference?

The shorthand is `{ headers: { "Authorization": "Bearer {YOUR-ENCODED-JWT}" } }`

If the `--fork.headers` option specifies an authorization header, it will be inserted after the JWT Bearer token.

**`--fork.userAgent`**

=== "Syntax"
    ```bash
    --fork.userAgent=<STRING>
    ```

=== "Example"

    ```bash
    --fork.userAgent="Ganache/7.0.0-beta.0 (https://www.trufflesuite.com/ganache; ganache<at>trufflesuite.com)]"
    ```

The User-Agent header sent to the fork on each request.

Sent as `Api-User-Agent` when used in the browser. Is overridden by a `User-Agent` defined in the `--fork.headers`
option, if provided.

> The default is `Ganache/7.0.0-beta.0 (https://www.trufflesuite.com/ganache; ganache<at>trufflesuite.com)]`

Reviewer: Aware this is coming from the README, but it feels a little funky (cc @davidmurdoch). Same goes for the example on [L745](https://github.com/trufflesuite/trufflesuite.com/pull/1312/files#diff-f929cf97d16f01f1875b744299cf3cd202190ad45781da6d4cffd10aed95bbbbR745).

Reviewer 2: The default value updates based off of the version of Ganache being run, so maybe we could update that accordingly? Also, there's an extra `]` at the end which could have been making it look more funky
```suggestion
The default is `Ganache/VERSION (https://www.trufflesuite.com/ganache; ganache<at>trufflesuite.com)`
```

---
title: Ganache | CLI options
layout: docs.hbs
---

# Ganache command line

This reference describes the syntax of the Ganache command line interface (CLI) options for the Ethereum
network.

## Specify options

The startup options are grouped in the `chain`, `database`, `fork`, `logging`, `miner`, and `wallet` namespaces,
and should be used as follows on startup:

```bash
ganache --<namespace>.<option>="<value>"
```

You can also use an alias for startup options that have them. For example to set the network ID of a
network use `ganache -i=<NETWORK_ID>` instead of `ganache --chain.networkId=<NETWORK_ID>`.

When using the startup options programmatically, use the following:

```javascript
const options = { <namespace>: { <option>: <"value">}};
const provider = ganache.provider(options);
```

<p class="alert alert-warning">
<strong>Note</strong>: Command line options are case-sensitive.
</p>

## `chain`

> **`--chain.allowUnlimitedContractSize`**

Author: Yes I think it would make it easier. However, that requires updates to the Website stylesheet, so would be best to implement in a separate PR.

    ```bash
    --miner.callGasLimit="0x58af080"
    ```

Set the transaction gas limit in WEI for `eth_call` and `eth_estimateGas` calls.

The default is `0x2faf080`.

**`--miner.instamine`**

=== "Syntax"
    ```bash
    --miner.instamine=<STRING>
    ```

=== "Example"

    ```bash
    --miner.instamine="strict"
    ```

Set the `instamine` mode to either `eager` or `strict`. In `eager` mode a transaction is included in a block before its hash
is returned to the caller. In `strict` mode a transaction's hash is returned to the caller before the transaction is included in a block.
This value has no effect if `--miner.blockTime` is *not* `0` (the default).

The default is `eager`.

**`--miner.coinbase`**

=== "Syntax"
    ```bash
    --miner.coinbase=<STRING>
    ```

=== "Example"

    ```bash
    --miner.coinbase="0xfe3b557e8fb62b89f4916b721be55ceb828dbd73"
    ```

Sets the mining reward address.

The default is `0x0000000000000000000000000000000000000000`.

**`--miner.extraData`**

=== "Syntax"
    ```bash
    --miner.extraData=<STRING>
    ```

=== "Example"

    ```bash
    --miner.extraData="0x444F4E27542050414E4943202120484F444C2C20484F444C2C20484F444C2021"
    ```

> A hex string representing the 32 bytes included in the extra data field of a mined block.

Reviewer 3: Until here, the option descriptions prescribe actions, since they start with "Set to..." / "Set the...". From this option forward, the descriptions omit "set" and only state the information required as arguments. We should use one or the other form for all the descriptions (I prefer the latter).

Author: This is what happens when you split work over two days :)

Yes I prefer the latter

The default is `0x`.

**`--miner.priceBump`**

=== "Syntax"
    ```bash
    --miner.priceBump=<STRING>
    ```

=== "Example"

    ```bash
    --miner.priceBump="0x1"
    ```

Minimum price bump percentage needed to replace a transaction that already exists in the transaction
pool.

The default is `10`.

> ## `wallet`

Author: These options follow the order when running `ganache --help`. But you make a good point, should we follow the order in the actual command, or alphabetically? We should discuss this, but for now I'll keep as is.

The alias `--accounts` has been deprecated, use `-a` instead.

**`--wallet.deterministic`**, **`-d`**

=== "Syntax"
    ```bash
    --wallet.deterministic=<BOOLEAN>
    ```

=== "Example"

    ```bash
    --wallet.deterministic=true
    ```

Use a pre-defined, deterministic seed.

The default is `false`.

The alias `--deterministic` has been deprecated, use `-d` instead.

**`--wallet.seed`**, **`-s`**

=== "Syntax"
    ```bash
    --wallet.seed=<STRING>
    ```

=== "Example"

    ```bash
    --wallet.seed="c0157f7c8240459516f42b518c1cba1da942b9a5fb5e6313a948872fcac2742e28250a8b2faef0924ae5d2d59781d7789c4c960a2fef8283d886198d40cf871f"
    ```

The seed to generate a mnemonic.

The default is a random value unless you specify `--wallet.deterministic`.

The alias `--seed` has been deprecated, use `-s` instead.

**`--wallet.mnemonic`**, **`-m`**

=== "Syntax"
    ```bash
    --wallet.mnemonic=<STRING>
    ```

=== "Example"

    ```bash
    --wallet.mnemonic="car casual program tourist aerobic broccoli link hamster resemble collect put october"
    ```

Use the specified HD wallet mnemonic to generate initial addresses.

The default is generated from `--wallet.seed`.

> The alias `--mnemonic` has been deprecated, use `-s` instead.

Reviewer 2: OOPS! My previous suggestion was wrong.
```suggestion
The alias `--mnemonic` has been deprecated, use `-m` instead.
```

Author: Fixed in the PR.

    ```

Specify the default account balance in ether.

The default is `1000`.

**`--wallet.hdPath `**

=== "Syntax"
    ```bash
    --wallet.hdPath=<STRING>
    ```

=== "Example"

    ```bash
    --wallet.hdPath="m,44',60',160720',0"
    ```

The hierarchical deterministic (HD) path to use when generating accounts.

The default is `m,44',60',0',0`.

## `fork`

**`--fork.url`**, **`-f`**

=== "Syntax"
    ```bash
    --fork.url=<STRING>
    ```

=== "Example"

    ```bash
    --fork.url="http://localhost:1337"
    ```

Fork from a running Ethereum client. You can optionally specify the block to
fork from using an `@` character, for example `http://localhost:1337@8675309`.

The specified URL supports basic authentication credentials as well. For example, `wss://<user>:<password>@example.com/`.
If you need to use an Infura API key secret, you would use it as follows:

```bash
wss://:<API-KEY-SECRET>@mainnet.infura.com/ws/v3/<API-KEY>
```

Alternatively, you can use the `--fork.username` and `--fork.password` options.

**`--fork.network`**

=== "Syntax"
    ```bash
    --fork.network=<STRING>
    ```

=== "Example"

    ```bash
    --fork.network="goerli"
    ```

> The network to fork. Valid options are `mainnet`, `ropsten`, `kovan`, `rinkeby`, `goerli`, `görli`,

Reviewer 2: Ganache v7.4.4 ware released yesterday, and it removes some testnets. We should use this new list:
mainnet, goerli, görli, sepolia

    ```

Headers to supply on each request to the forked provider.
Headers set here override headers set by other options, unless otherwise specified.

Defaults to `"User-Agent: Ganache/VERSION (https://www.trufflesuite.com/ganache; ganache<at>trufflesuite.com)"`.

**`--fork.requestsPerSecond`**

=== "Syntax"
    ```bash
    --fork.requestsPerSecond=<NUMBER>
    ```

=== "Example"

    ```bash
    --fork.requestsPerSecond=10
    ```

Restrict the number of requests that are sent per second to the fork provider.

The default is `0`, meaning no limit is applied.

**`--fork.disableCache`**

=== "Syntax"
    ```bash
    --fork.disableCache=<BOOLEAN>
    ```

=== "Example"

    ```bash
    --fork.disableCache=true
    ```

Disables caching forking requests.

The default is `false`.

**`--fork.deleteCache`**

=== "Syntax"
    ```bash
    --fork.deleteCache=<BOOLEAN>
    ```

=== "Example"

    ```bash
    --fork.deleteCache=true
    ```

Deletes the persistent cache before starting.

The default is `false`.

## `server`

**`--server.ws`**

=== "Syntax"
    ```bash
    --server.ws=<BOOLEAN>
    ```

=== "Example"

    ```bash
    --server.ws=false
    ```

> Enable a websocket server.

Reviewer 3: ```suggestion
Enables a WebSocket server.
```

The default is `true`.

**`--server.wsBinary`**

=== "Syntax"
    ```bash
    --server.wsBinary=<BOOLEAN>
    ```

=== "Example"

    ```bash
    --server.wsBinary=false
    ```

> Whether or not Websockets should respond with binary data (ArrayBuffers) or strings. Options are

Reviewer 3: ```suggestion
Indicates whether WebSocket should respond with binary data (ArrayBuffers) or strings. Options are
```

**`--miner.timestampIncrement`**

=== "Syntax"

    ```bash
    --miner.timestampIncrement=<NUMBER>
    ```

=== "Example"

    ```bash
    --miner.timestampIncrement=5
    ```

The amount of time (in seconds) to add to the timestamp of each new block header.

The default is `clock`, which uses your system clock time as the timestamp for each new block.

**`--miner.defaultGasPrice`**, **`g`**

=== "Syntax"

    ```bash
    --miner.defaultGasPrice=<STRING>
    ```

=== "Example"

    ```bash
    --miner.defaultGasPrice="0x87369400"
    ```

The default gas price in WEI for transactions if not specified.

The default is `0x77359400` (2 GWEI).

The alias `--gasPrice` has been deprecated, use `-g` instead.

**`--miner.blockGasLimit`**, **`l`**

=== "Syntax"
    ```bash
    --miner.blockGasLimit=<STRING>
    ```

=== "Example"

    ```bash
    --miner.blockGasLimit="0x87369400"
    ```

The block gas limit in WEI.

The default is `0x1c9c380` (30 million WEI).

The alias `--gasLimit` has been deprecated, use `-l` instead.

**`--miner.defaultTransactionGasLimit`**

=== "Syntax"
    ```bash
    --miner.defaultTransactionGasLimit=<STRING>
    ```

=== "Example"

    ```bash

>     --miner.defaultTransactionGasLimit="16F30"

Reviewer 2: ```suggestion
    --miner.defaultTransactionGasLimit="0x16F30"
```
